### PR TITLE
resolve: cleanup and groundwork for resolving the AST

### PR DESCRIPTION
**Pull request by jseyfried**
Cleanup `resolve` and refactor away uses of the hir map (incorrectly named `ast_map`).
r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @petrochenkov

File: src/librustc/ty/mod.rs [hunk 1]
```diff
@@ -283,6 +283,22 @@ pub enum Visibility {
...
+    fn is_descendant_of(&self, node: NodeId, ancestor: NodeId) -> bool;
+}
+
+impl<'a> NodeIdTree for ast_map::Map<'a> {
+    fn is_descendant_of(&self, node: NodeId, ancestor: NodeId) -> bool {
+        let mut node_ancestor = node;
+        loop {
+            if node_ancestor == ancestor { return true }
+            let node_ancestor_parent = self.get_module_parent(node_ancestor);
+            if node_ancestor_parent == node_ancestor { return false }
```
Comment (nrc) [inline src/librustc/ty/mod.rs]: style nit: spread these `if`s over three lines

Comment (jseyfried) [inline src/librustc/ty/mod.rs]: Will do

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -1135,10 +1121,19 @@ impl<'a> ResolverArenas<'a> {
...
-#[derive(PartialEq)]
-enum FallbackChecks {
-    Everything,
-    OnlyTraitAndStatics,
+impl<'a, 'tcx> ty::NodeIdTree for Resolver<'a, 'tcx> {
+    fn is_descendant_of(&self, node: NodeId, ancestor: NodeId) -> bool {
+        let ancestor = self.ast_map.local_def_id(ancestor);
+        let mut module = *self.module_map.get(&node).unwrap();
+        loop {
+            if module.def_id() == Some(ancestor) { return true; }
```
Comment (nrc) [inline src/librustc_resolve/lib.rs]: here too

File: src/librustc_resolve/lib.rs [hunk 3]
```diff
@@ -1030,7 +1016,7 @@ pub struct Resolver<'a, 'tcx: 'a> {
 
     graph_root: Module<'a>,
 
-    trait_item_map: FnvHashMap<(Name, DefId), DefId>,
+    trait_item_map: FnvHashMap<(Name, DefId), bool /* is static method? */>,
```
Comment (nrc) [inline src/librustc_resolve/lib.rs]: I'd prefer to have a Static/Instance (or something) enum here, but I don't feel too strongly.

**Comment by nrc** [comment]
Nice! r+ with the style nits addressed.

**Comment by jseyfried** [comment]
@nrc thanks for the quick review! I addressed the style nits in the above commit.

File: src/librustc_resolve/lib.rs [hunk 4]
```diff
@@ -1125,14 +1125,14 @@ impl<'a, 'tcx> ty::NodeIdTree for Resolver<'a, 'tcx> {
...
-        loop {
-            if module.def_id() == Some(ancestor) { return true; }
+        while module.def_id() != Some(ancestor) {
             let module_parent = match self.get_nearest_normal_module_parent(module) {
                 Some(parent) => parent,
                 None => return false,
             };
             module = module_parent;
         }
+        return true;
```
Comment (nrc) [inline src/librustc_resolve/lib.rs]: prefer `true` rather than `return true;`

Comment (nrc) [inline src/librustc_resolve/lib.rs]: and in the other function too

**Comment by jseyfried** [comment]
@nrc
> prefer `true` rather than `return true;`

Of course -- didn't notice that when mechanically refactoring :)
Fixed.

**Pull request merged**